main/lsp/requests/completion.cc

```diff
@@ -262,8 +277,52 @@ unique_ptr<CompletionItem> getCompletionItemForKeyword(const core::GlobalState &
     return item;
 }
 
+unique_ptr<CompletionItem> getCompletionItemForLocal(const core::GlobalState &gs, const LSPConfiguration &config,
+                                                     const core::LocalVariable &local, const core::Loc queryLoc,
+                                                     string_view prefix, size_t sortIdx) {
+    auto label = string(local._name.data(gs)->shortName(gs));
+    auto item = make_unique<CompletionItem>(label);
+    item->sortText = fmt::format("{:06d}", sortIdx);
+    item->kind = CompletionItemKind::Variable;
+
+    // TODO(jez) This should probably be a helper function (see getCompletionItemForSymbol)
+    u4 queryStart = queryLoc.beginPos();
+    u4 prefixSize = prefix.size();
+    auto replacementLoc = core::Loc{queryLoc.file(), queryStart - prefixSize, queryStart};
+    auto replacementRange = Range::fromLoc(gs, replacementLoc);
+    auto replacementText = label;
+    if (replacementRange != nullptr) {
+        item->textEdit = make_unique<TextEdit>(std::move(replacementRange), replacementText);
+    } else {
+        // TODO(jez) Why is replacementRange nullptr? instrument this and investigate when it fails
+        item->insertText = replacementText;
+    }
+    item->insertTextFormat = InsertTextFormat::PlainText;
+    // TODO(jez) Show the type of the local under the documentation field?
```

Comment:
> If computing full completion items is expensive, servers can additionally provide a handler for the completion item resolve request (‘completionItem/resolve’). This request is sent when a completion item is selected in the user interface. A typical use case is for example: the ‘textDocument/completion’ request **doesn’t fill in the documentation property for returned completion items since it is expensive to compute**. When the item is selected in the user interface then a ‘completionItem/resolve’ request is sent with the selected completion item as a parameter. The returned completion item should have the documentation property filled in. The request can delay the computation of the detail and documentation properties.

https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#textDocument_completion

I found the language about the feature we discussed IRL. Seems like the right solution to me.

Reply:
I just realized that this might be hard because no existing query we can send will return the type information. We'd have to change `infer` to support something like this.